README.md
# AppScan on Cloud AutoScanner

This script will compress a directory into a zip file and upload to HCL AppScan on Cloud for Static Analysis.

## Environment Variables

Environment variables are used to configure the script. They are specified below.

| Env Var Name  | Description | Required |
| ------------- | ------------- | ------------- |
| ASOC_ORG_NAME  | The name of the organization in ASoC  | No |
| ASOC_PROJECT_NAME  | The name of the project to scan. This will be created in ASoC if it does not exist  | Yes |
| ASOC_TARGET_DIR  | The full path to the directory to be zipped and scanned | Yes |
| ASOC_API_KEY_ID  | The AppScan on Cloud API Key ID  | Yes |
| ASOC_API_KEY_SECRET  | The AppScan on Cloud API Key Secret  | Yes |

Added support for HTTP_PROXY and HTTPS_PROXY environment variables. 

Example: `HTTPS_PROXY=123.123.123.123:4444`

AppScan on Cloud API Key may be generated inside the AppScan on Cloud portal via the menu Tools > API > Generate.

## Running a scan

To execute a scan with this script the above env variables will need to be created. Python and Pip are needed as well. This script was built and tested with Python 3.11.0.

Clone the repo
```bash
git clone https://github.com/cwtravis/asoc_autoscanner.git
cd asoc_autoscanner
```

Install the requirements via Pip
```bash
python3 -m pip install -r requirements.txt
```

Run the scan
```bash
python3 run_sast.py
```


CHANGELOG.md
# ASoC AutoScanner Changelog

All notable changes to this project will be documented in this file.

## Version 1.0.1 - 2023-07-07

- Added and allow list for file extensions and files to limit the contents of the zip file to only those the scanner cares about.
- Added an appscan-config.xml file to the zip to enforce scan preferences.
- Adjusted ASoC API calls to only includes neccessary fields to speed them up.

## Version 1.0.1 - 2023-07-06

- Added support for environment variables "HTTP_PROXY" and "HTTPS_PROXY". These will be picked up by the scripts and honored.
- - Example:  `HTTPS_PROXY=123.123.123.123:4444`
- ASoC object now uses a requests Session object to manage persistent proxy and header data

## Version 1.0.0 - Initial Version

- Released

asoc.py
import requests
import time
import datetime
import os

class ASoC:
    def __init__(self, apikey_id, apikey_secret, proxies=None):
        self.apikey = {
            "KeyId": apikey_id,
            "KeySecret": apikey_secret
        }
        self.token = ""
        self.session = requests.Session()
        if proxies is not None:
            self.session.proxies.update(proxies)
        self.session.headers.update({"Accept": "application/json"})
    
    def login(self):
        resp = self.session.post("https://cloud.appscan.com/api/V2/Account/ApiKeyLogin", json=self.apikey)
        if(resp.status_code == 200):
            jsonObj = resp.json()
            self.token = jsonObj["Token"]
            self.session.headers.update({"Authorization": "Bearer "+self.token})
            return True
        else:
            print(f"Status Code: {resp.status_code}")
            print(resp.text)
            return False
        
    def logout(self):
        resp = self.session.get("https://cloud.appscan.com/api/V2/Account/Logout")
        if(resp.status_code == 200):
            self.token = ""
            return True
        else:
            return False
        
    def checkAuth(self):
        resp = self.session.get("https://cloud.appscan.com/api/V2/Account/TenantInfo")
        return resp.status_code == 200

    def getApps(self, name=None):
        json_obj = None
        data = {
            "$filter": f"Name eq '{name}'"
        }
        if name is not None:
            resp = self.session.get(f"https://cloud.appscan.com/api/V2/Apps/GetAsPage?$select=Id", params=data)
        else:
            resp = self.session.get(f"https://cloud.appscan.com/api/V2/Apps/GetAsPage?$select=Id")
        if resp.status_code >= 200 and resp.status_code < 400:
            json_obj = resp.json()
        return (resp.status_code, json_obj)
    
    def getAssetGroup(self, name=None, filter_default=False):
        json_obj = None
        if name is not None:
            filter_str = f"name eq '{name}'" 
        else:
            filter_str = ""
        if filter_default:
            if len(filter_str)>0:
                filter_str += " and "
            filter_str += "IsDefault eq true"
        data = {
            "$filter": filter_str
        }
        resp = self.session.get("https://cloud.appscan.com/api/V2/AssetGroups", params=data)
        if resp.status_code >= 200 and resp.status_code < 400:
            json_obj = resp.json()
        return (resp.status_code, json_obj)

    def createApp(self, name, asset_group_id):
        app_info = {
            "Name": name,
            "AssetGroupId": asset_group_id
        }
        resp = self.session.post("https://cloud.appscan.com/api/V2/Apps", json=app_info)
        return resp.status_code, resp.json()

    def uploadFile(self, file_path):
        file_name = os.path.basename(file_path)
        files = {"fileToUpload":(file_name, open(file_path, 'rb'), 'application/x-zip-compressed')}
        resp = self.session.post("https://cloud.appscan.com/api/V2/FileUpload?fileType=SourceCodeArchive", files=files)
        return resp.status_code, resp.json()

    def sastScan(self, file_id, app_id, scan_name="SAST Scan"):
        data = {
            "ApplicationFileId": file_id,
            "ScanName": scan_name,
            "EnableMailNotification": False,
            "AppId": app_id
        }
        resp = self.session.post("https://cloud.appscan.com/api/V2/Scans/StaticAnalyzer", data=data)
        return resp.status_code, resp.json()

    def scanDetails(self, scan_id):
        resp = self.session.get(f"https://cloud.appscan.com/api/V2/Scans/{scan_id}")
        return resp.status_code, resp.json()
    
    def scanStatus(self, scan_id):
        resp = self.session.get(f"https://cloud.appscan.com/api/V2/Scans/{scan_id}")
        if resp.status_code == 403:
            return "Cancelled"
        if resp.status_code >= 300:
            return "Error"
        json_obj = resp.json()
        return json_obj["LatestExecution"]["Status"]
    
    @staticmethod
    #Get current system timestamp
    def getTimeStamp():
        ts = time.time()
        return datetime.datetime.fromtimestamp(ts).strftime('%Y-%m-%d_%H-%M-%S')
    
ALLOW_LIST_EXTENSIONS = [
    ".java",
    ".cls",
    ".page",
    ".c",
    ".cpp",
    ".mm",
    ".h",
    ".h++",
    ".hh",
    ".hxx",
    ".hpp",
    ".H",
    ".ii",
    ".ixx",
    ".ipp",
    ".inl",
    ".txx",
    ".tpp",
    ".tpl",
    ".c+",
    ".cc",
    ".cxx",
    ".C",
    ".cob",
    ".cbl",
    ".ws",
    ".sqb",
    ".cfc",
    ".cfm",
    ".dart",
    ".bat",
    ".sh",
    ".yaml",
    ".yml",
    ".tf",
    ".tf.json",
    ".properties",
    ".curl",
    ".ini",
    ".conf",
    ".js",
    ".hbs",
    ".htm",
    ".html",
    ".rhtml",
    ".xhtml",
    ".cshtml",
    ".vbhtml",
    ".jsf",
    ".jsx",
    ".jsp",
    ".jspx",
    ".jspi",
    ".aspx",
    ".ascx",
    ".asp",
    ".asax",
    ".asa",
    ".inc",
    ".php",
    ".php3",
    ".php4",
    ".php5",
    ".phtm",
    ".phps",
    ".rjs",
    ".wlapp",
    ".ts",
    ".tsx",
    ".svg",
    ".vue",
    ".kt",
    ".m",
    ".mm",
    ".java",
    ".jsp",
    ".jspx",
    ".jspf",
    ".pl",
    ".pm",
    ".cgi",
    ".t",
    ".ctp",
    ".html",
    ".php",
    ".php4",
    ".php3",
    ".php5",
    ".php6",
    ".php7",
    ".phtm",
    ".phps",
    ".inc",
    ".htaccess",
    ".module",
    ".yaml",
    ".yml",
    ".xml",
    ".py",
    ".pyt",
    ".pyw",
    ".rpg",
    ".rpgl",
    ".rpgle",
    ".rb",
    ".rhtml",
    ".rjs",
    ".erb",
    ".abap",
    ".swift",
    ".plist",
    ".sql",
    ".lst",
    ".dbf",
    ".rdo",
    ".arc",
    ".pls",
    ".plb",
    ".pks",
    ".pkb",
    ".pck",
    ".tst",
    ".sp",
    ".sf",
    ".spb",
    ".sps",
    ".sql",
    ".lst",
    ".dbf",
    ".rdo",
    ".arc",
    ".dll",
    ".exe",
    ".asp",
    ".asa",
    ".inc",
    ".bas",
    ".cls",
    ".frm",
    ".jsp",
    ".jspx",
    ".jspi",
    ".htm",
    ".html",
    ".xml",
    ".java",
    ".class",
    ".jar",
    ".wsdl",
    ".smap",
    ".tld",
    ".tag",
    ".war",
    ".ear",
    ".rar",
    ".js",
    ".cpp",
    ".c",
    ".C",
    ".cc",
    ".cxx",
    ".CPP",
    ".CC",
    ".CXX",
    ".h",
    ".hpp",
    ".H",
    ".HPP",
    ".HH",
    ".hh",
    ".i",
    ".inl",
    ".txt",
    ".sql",
    ".ico",
    ".rc",
    ".ICO",
    ".RC",
    ".cur",
    ".rc2",
    ".RC2",
    ".bmp",
    ".aspx",
    ".cs",
    ".vb",
    ".vjs",
    ".master",
    ".xml",
    ".ico",
    ".config",
    ".xsd",
    ".xss",
    ".sql",
    ".mdf",
    ".ldf",
    ".ashx",
    ".css",
    ".ascx",
    ".htm",
    ".txt",
    ".html",
    ".asax",
    ".gif",
    ".jpg",
    ".sitemap",
    ".bmp",
    ".js",
    ".dll",
    ".disco",
    ".vsdisco",
    ".wsdl",
    ".map",
    ".resx",
    ".png",
    ".mdb",
    ".cshtml",
    ".vbhtml",
    ".aspx",
    ".cs",
    ".vb",
    ".vjs",
    ".master",
    ".xml",
    ".ico",
    ".config",
    ".xsd",
    ".xss",
    ".sql",
    ".mdf",
    ".ldf",
    ".ashx",
    ".css",
    ".ascx",
    ".htm",
    ".txt",
    ".html",
    ".asax",
    ".gif",
    ".jpg",
    ".sitemap",
    ".bmp",
    ".js",
    ".dll",
    ".disco",
    ".vsdisco",
    ".wsdl",
    ".map",
    ".resx",
    ".png",
    ".mdb",
    ".cshtml",
    ".vbhtml",
    ".aspx",
    ".ascx",
    ".cs",
    ".cs",
    ".vb",
    ".vbs",
    ".vb",
    ".vbs",
    ".wsdl",
    ".xsd",
    ".go",
    ".scala",
    ".sc",
    ".groovy",
    ".gsh",
    ".gsp",
    ".gvy",
    ".gy",
    ".ts",
    ".tsx",
    ".cs",
    ".rs",
    ".json",
    ".toml",
    ".json5"
]

ALLOW_LIST_FILES = [
    "androidmanifest.xml",
    "web.config",
    "dockerfile",
    "docker-compose",
    "config.xml",
    "androidmanifest.xml",
    "php.ini",
    "httpd.conf"
]

run_sast.py
import os
import sys
import zipfile
import time
from asoc import ASoC, ALLOW_LIST_EXTENSIONS, ALLOW_LIST_FILES

start_time = time.time()

required_args = [
    "ASOC_PROJECT_NAME",
    "ASOC_TARGET_DIR",
    "ASOC_API_KEY_ID",
    "ASOC_API_KEY_SECRET",
]

# Check Required Env Vars Exists
for arg in required_args:
    if arg not in os.environ:
        print(f"Required env variable not found: {arg}")
        print("Please specify all required env vars. They are:\n"+", ".join(required_args))
        sys.exit(1)

if "ASOC_ORG_NAME" not in os.environ.keys():
    org_name = "Not Specified"
else:
    org_name = os.environ["ASOC_ORG_NAME"]
project_name = os.environ["ASOC_PROJECT_NAME"]
directory = os.environ["ASOC_TARGET_DIR"]
api_key_id = os.environ["ASOC_API_KEY_ID"]
api_key_secret = os.environ["ASOC_API_KEY_SECRET"]
app_id = None

# Check for proxies
proxies = {}
http_proxy = None
https_proxy = None
if "HTTP_PROXY" in os.environ.keys():
    http_proxy = os.environ["HTTP_PROXY"]
    if "http://" not in http_proxy:
        http_proxy = f"http://{http_proxy}"
        proxies["http"] = http_proxy

if "HTTPS_PROXY" in os.environ.keys():
    https_proxy = os.environ["HTTPS_PROXY"]
    if "https://" not in https_proxy:
        https_proxy = f"https://{https_proxy}"
        proxies["https"] = https_proxy

if len(proxies.keys()) == 0:
    proxies = None
else:
    print("Proxy Info:")
    print(proxies)


print("HCL AppScan on Cloud - Run SAST")
print("------------------------------")
print(f"Org Name: {org_name}")
print(f"Project Name: {project_name}")
print(f"Target Directory: {directory}")
print(f"ASOC API Key: **provided**")

default_scan_name = project_name + "_" + ASoC.getTimeStamp()
scantarget_zip = default_scan_name + ".zip"
cwd = cwd = os.getcwd() 
scantarget_zip_path = os.path.join(cwd, scantarget_zip)

print(f"Zip File: {scantarget_zip_path}")
print("------------------------------")
app_check_time = time.time()
# Check Project Name exists as App
asoc = ASoC(api_key_id, api_key_secret, proxies=proxies)
if asoc.login():
    code, json_obj = asoc.getApps(project_name)
    if code != 200:
        print(f"ASoC API Response Code: {code}")
        print("Could not query ASoC API for App Info")
        sys.exit(1)
    if len(json_obj['Items']) > 0:
        # App Exists Already
        app_id = json_obj['Items'][0]["Id"]
        print(f"App [{project_name}] Exists: [{app_id}]")
    else:
        # Need to create the app
        # ToDo: Find AssetGroupId of Default Asset Group
        code, json_obj = asoc.getAssetGroup(filter_default=True)
        if code != 200:
            print("Could not get Default AssetGroup Id")
            sys.exit(1)
        default_asset_group_id = json_obj[0]["Id"]
        code, json_obj = asoc.createApp(project_name, default_asset_group_id)
        if code >= 300:
            print(f"Could not create application in ASoC: {project_name} in asset_group {default_asset_group_id}")
            print(f"Invalid Status Code: {code}")
            print(json_obj)
            sys.exit(1)
        app_id = json_obj['Id']
        print(f"Created App [{project_name} - {app_id}]")
else:
    print("ASoC API Key Login was unsuccessful. Bad API key?")
    sys.exit(1)

# Check Target Directory Exists
if not os.path.isdir(directory):
    print(f"Target Direcotry {directory} does not exist.")
    print("Create the directory or specify one that exists.")
    sys.exit(1)

app_check_time = round(time.time()-app_check_time)

zip_time = time.time()
print("Zipping Target Directory")
with zipfile.ZipFile(scantarget_zip, "w", zipfile.ZIP_DEFLATED) as zipf:
    zipf.write('appscan-config.xml')
    for root, dirs, files in os.walk(directory):
        for file in files:
            f_name, f_ext = os.path.splitext(file)
            if file.lower() in ALLOW_LIST_FILES or f_ext in ALLOW_LIST_EXTENSIONS:
                zipf.write(os.path.join(root, file), 
                    os.path.relpath(os.path.join(root, file), 
                    os.path.join(directory, '..')))
print("Zip Complete")    
file_size = os.path.getsize(scantarget_zip)
file_size_kb = file_size / 1024
file_size_mb = file_size_kb / 1024
print(f"Zip File Size: {round(file_size_mb)} MB")
zip_time = round(time.time()-zip_time)

print("------------------------------")

upload_time = time.time()
print("Uploading zip to ASoC")
code, json_obj = asoc.uploadFile(scantarget_zip)
if code >= 300:
    print("Error uploading zip file.")
    print(f"Invalid Response Code: {code}")
    print(json_obj)
    sys.exit(1)

file_id = json_obj["FileId"]
print(f"Zip file uploaded: FileId [{file_id}]")
upload_time = round(time.time()-upload_time)

print("------------------------------")
scan_time = time.time()

print("Creating SAST Scan")
code, json_obj = asoc.sastScan(file_id, app_id, scantarget_zip)
if code >= 300:
    print("Error creating scan")
    print(f"Invalid Response Code: {code}")
    print(json_obj)
    sys.exit(1)

scan_id = json_obj["Id"]
print(f"Scan Created: Id [{scan_id}]")
print("Waiting for Scan to complete...")

# Wait for scan to complete
old_status = ""
status = asoc.scanStatus(scan_id)
error_count = 0
while status not in ["Paused", "Ready", "Failed"]:
    if old_status != status :
        print(f"\tCurrent Status: {status}")
        old_status = status
    if status == "Error":
        error_count += 1
    else:
        error_count = 0
    if error_count >= 6:
        print("Error attempting to get the current scan status")
        sys.exit(1)
    if status == "Cancelled":
        print("API Returned Status Code 403 - Scan may have been cancelled.")
        sys.exit(0)
    time.sleep(15)
    status = asoc.scanStatus(scan_id)

print("Scan Complete")
code, json_obj = asoc.scanDetails(scan_id)
if code >= 300:
    print("Error getting scan details")
    print(json_obj)
    sys.exit(1)
scan_time = round(time.time()-scan_time)
print("------------------------------")

NIssuesFound = json_obj["LatestExecution"]["NIssuesFound"]
NCriticalIssues = json_obj["LatestExecution"]["NCriticalIssues"]
NHighIssues = json_obj["LatestExecution"]["NHighIssues"]
NMediumIssues = json_obj["LatestExecution"]["NMediumIssues"]
NLowIssues = json_obj["LatestExecution"]["NLowIssues"]
NInfoIssues = json_obj["LatestExecution"]["NInfoIssues"]

print("Scan Summary:")
print(f"\tCritical: {NCriticalIssues}")
print(f"\tHigh: {NHighIssues}")
print(f"\tMed: {NMediumIssues}")
print(f"\tLow: {NLowIssues}")
print(f"\tInfo: {NInfoIssues}")
print()
print(f"Total Issues: {NIssuesFound}")

# Logout of ASoC
asoc.logout()

elapsed_time = round(time.time()-start_time)
print(f"Runtime Summary: ")
print(f"\t App Check Time: {app_check_time} sec")
print(f"\t Zip Time: {zip_time} sec")
print(f"\t Upload Time: {upload_time} sec")
print(f"\t Scan Time: {scan_time} sec")
print(f"\t Total Time: {elapsed_time} sec")